packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,153 @@
+    }
+
+    return value.map((y) => {
+      return ts2aqua(y, schema.type);
+    });
+  } else if (schema.tag === "unlabeledProduct") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
 
-  if (arrowType.codomain.items.length === 1) {
-    return ts2aqua(returnValue, arrowType.codomain.items[0]);
+    return value.map((y, i) => {
+      return ts2aqua(y, schema.items[i]);
+    });
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = ts2aqua(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
+}
+
+export const wrapFunction = (
+  value: (...args: JSONValue[]) => JSONValue,
+  schema: ArrowWithoutCallbacks,
+): ((...args: JSONValue[]) => JSONValue) => {
+  return (...args) => {
+    const schemaArgs =
+      schema.codomain.tag === "nil" ? [] : schema.codomain.items;
+
+    const tsArgs = args.map((arg, i) => {
+      return aqua2ts(arg, schemaArgs[i]);
+    });
 
-  return arrowType.codomain.items.map((type, index) => {
-    return ts2aqua(returnValue[index], type);
-  });
+    const result = value(...tsArgs);
+    return ts2aqua(result, schema.codomain);
+  };
 };
 
-/**
- * Converts response value from aqua its representation to representation in typescript
- * @param req - call service data
- * @param arrow - aqua type definition
- * @returns response value in typescript representation
- */
-export const responseServiceValue2ts = (
-  req: CallServiceData,
-  arrow: ArrowType<any>,
-) => {
-  return match(arrow.codomain)
-    .with({ tag: "nil" }, () => {
-      return null;
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      if (x.items.length === 0) {
-        return null;
-      }
-
-      if (x.items.length === 1) {
-        return aqua2ts(req.args[0], x.items[0]);
-      }
-
-      return req.args.map((y, index) => {
-        return aqua2ts(y, x.items[index]);
-      });
-    })
-    .exhaustive();
+export const wrapServiceFunction = (
+  value: (
+    ...args: [...JSONValue[], ParticleContext]
+  ) => MaybePromise<JSONValue>,
+  schema: ArrowWithCallbacks,
+): ((
+  ...args: [...JSONValue[], ParticleContext]
+) => MaybePromise<JSONValue>) => {
+  return async (...args) => {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const jsonArgs = args.slice(0, args.length - 1) as JSONValue[];
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const context = args[args.length - 1] as ParticleContext;
+
+    const schemaArgs =
+      schema.codomain.tag === "nil" ? [] : schema.codomain.items;
+
+    const tsArgs = jsonArgs.map((arg, i) => {
```

Comment:
doc outlining intention here would be nice

Reply:
no docs here were added as well, also if you renamed the function to `aqua2js` - maybe the variable has to be renamed as well here